common/predictive-text/worker/correction/context-tracker.ts

```diff
@@ -14,16 +14,93 @@ namespace correction {
   }
 
   export class TrackedContextState {
-    context: Context;
+    // Stores the source Context (as a debugging reference).  Not currently utilized.
+    taggedContext: Context;
+
     tokens: TrackedContextToken[];
-    poppedHead: boolean;
-    pushedTail: boolean;
-    
+    indexOffset: number;
+
     // Tracks all search spaces starting at the current token.
     // In the lm-layer's current form, this should only ever have one entry.
     // Leaves 'design space' for if/when we add support for phrase-level corrections/predictions.
     searchSpace: SearchSpace[] = [];
 
+    constructor(source: TrackedContextState);
+    constructor(rootTraversal: LexiconTraversal);
+    constructor(obj: TrackedContextState | LexiconTraversal) {
+      if(obj instanceof TrackedContextState) {
+        let source = obj;
+        // Be sure to deep-copy the tokens!  Pointer-aliasing is bad here.
+        this.tokens = source.tokens.map(function(token) {
+          let copy = new TrackedContextToken();
+          copy.raw = token.raw;
+          copy.replacements = token.replacements
+          copy.activeReplacement = token.activeReplacement;
+          copy.transformDistributions = token.transformDistributions;
+  
+          return copy;
+        });
+        this.searchSpace = obj.searchSpace;
+        this.indexOffset = 0;
+      } else {
+        let rootTraversal = obj;
+        this.tokens = [];
+        this.indexOffset = Number.MIN_VALUE;
```

Comment:
I don't know if this matters or not, but just note that `Number.isInteger(Number.MIN_VALUE)` is `false` (e.g., `Number.MIN_VALUE` is *not* an integer).

If you just want something that will always compare as less than to everything (except `NaN` 🙄 ), use `-Infinity`.

Reply:
According to Firefox... `-Infinity` returns `false`, too.  That said, _both_ of them return `false` from `Number.isNaN`, so they both are counted as numbers.  That said... yeah, the `<` comparison doesn't seem to work correctly then - good point.